substrate/client/consensus/beefy/src/worker.rs

```diff
@@ -810,6 +830,35 @@ where
 		}
 	}
 
+	/// execute the ETF extract algorithm
+	/// outputs a (threshold) IBE secret and corresponding DLEQ proof
+	fn etf_extract(
+		&mut self,
+		hash: B::Hash,
+		id: AuthorityId,
+		message: &[u8]
+	) -> Option<Signature> {
+		let runtime_api = self.runtime.runtime_api();
+		info!(
+			target: LOG_TARGET,
+			"🎲 run ACSS recovery at best grandpa: #{:?}.",
+			hash
+		);
+		if let Some(Some(validator_set)) = runtime_api.validator_set(hash).ok() {
+			if let Some(Some(pok_bytes)) = runtime_api.read_share(hash, id.clone()).ok() {
+				if let Ok(sig)  = self.key_store.etf_sign(
+					&id, 
+					&pok_bytes, 
+					&message, validator_set.len() as u8
+				) {
+					return Some(sig);
+				}
+			}
+		}
```

Comment:
Optional: I usually find `match` statements easier to read than nested `if`s

```suggestion
		match (runtime_api.validator_set(hash), runtime_api.read_share(hash, id.clone()) {
			(Some(validator_set), Some(pok_bytes)) if let Ok(sig)  = self.key_store.etf_sign(
				&id, 
				&pok_bytes, 
				&message, validator_set.len() as u8
			) => return Some(sig);
			_ => return None;
		}
```